### DISCUSSION TITLE
MCP23017 not showing up

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22410] [Q&A]
hello. I start playing with MCP23017, target ist to use mor than one.
But Idon't get it start working. 

actually i only add these for compiling:
```
#define USE_MCP23XXX_DRV
#define MCP23XXX_ADDR_START      0x20     // 32
#define MCP23XXX_ADDR_END        0x26     // 38
#define MCP23XXX_MAX_DEVICES     6
```

mcp23x.dat (in filesystem):
```
{"NAME":"MCP23017 expander","BASE":0,"GPIO":[224,225,226,227,32,33,34,35]
```

cosole:
```
...
17:45:47.926 RSL: INFO3 = {"Info3":{"RestartReason":"RTC Watch dog reset digital core and rtc module","BootCount":213}}
17:45:47.939 RSL: RESULT = {"POWER":"ON"}
17:45:47.939 RSL: POWER = ON
17:45:50.909 QPC: Reset
17:45:51.901 RSL: STATE = {"Time":"2024-11-01T17:45:51","Uptime":"0T00:00:08","UptimeSec":8,"Heap":174,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":7,"Objects":76},"POWER":"ON","Wifi":{"AP":1,"SSId":"xxxxxxxxx","BSSId":"xxxxxxxxxxx,"Channel":1,"Mode":"HT40","RSSI":82,"Signal":-59,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:45:53.491 CMD: i2cscan
17:45:53.509 RSL: RESULT = {"I2CScan":"Device(s) found on bus1 at 0x26 0x5d"}
17:45:55.746 CMD: i2cdriver
17:45:55.748 RSL: RESULT = {"I2CDriver":"7,9,10,58,59,77"}
```
![grafik](https://github.com/user-attachments/assets/8ae163c5-9ec6-4fb9-8a07-04b63b92b949)

**Comment by sfromis** [comment]
Except for `#define USE_MCP23XXX_DRV`, your `#define` are not relevant to mode 2, and as the code is implemented, it expects your MCP23017 devices to be configured with I2C addresses starting with `0x20` up to `0x25`, hence not `0x26`.

**Reply by ChrisVonFrosta** [reply]
Thank you!
but i do not understand exaclty.
If the defines do not effect to mode 2 than my module with 0x26 should be detected ?
To be sure, i compiled again with:
```
#define USE_MCP23XXX_DRV
#define MCP23XXX_ADDR_START      0x20     // 32
#define MCP23XXX_ADDR_END        0x27     // 38
#define MCP23XXX_MAX_DEVICES     6
```

what didn't help. What you just explained - not effected to mode 2.

**Reply by sfromis** [reply]
Well, I very *explicitly* told you that those `#define` (except for the first) are irrelevant, and that the current implementation of that driver is **only** up to `0x25`.

(Not using those expander chips, I'm not trying to suggest what's "right", docs vs driver, only observe how it currently works)

**Reply by ChrisVonFrosta** [reply]
ah ok, than i miss understand the last thing about the adress range.i understand you said some:
"if #define MCP23XXX_ADDR_END  0x27  would work, it would means, last adress is 0x26."

Thankyou for quick your response!

**Reply by ChrisVonFrosta** [reply]
the device came default wth 0x027 - no jumper soldered.
is there any reason, why i should not expand the adress range in the `xdrv_67_mcp23xxx.ino`?

why do `#define MCP23XXX_ADDR_END        0x26     // 38` in `user_config_override.h` not redfine the the value in 
`xdrv_67_mcp23xxx.ino` for mode 2?

**Reply by barbudor** [reply]
In Tasmota 0x27 is reserved for i2x Lcd Displays which generally cannot get another address, while your module can be changed
So no, the default end address will not be changed to 0x27 as you can make your own build or solder pads. It has to remains to insure max compatibility.

**Reply by sfromis** [reply]
And once more, with current setup for the mode 2 driver, the last address scanned is `0x25`, not `0x26`. Arguably it would not need to be this way (while still not scanning `0x27` due to the reasons explained by barbudor), but that's the current implementation.

You cannot override `MCP23XXX_ADDR_END` as the `#define` of this symbol in the driver is not wrapped in `#ifndef` / `#endif`. Thus, only if you modify the driver code. It would not be unreasonable to do such a modification to the standard driver, to allow a build time upgrade.

**Reply by ChrisVonFrosta** [reply]
thx for clearance.

i was confused,because the `#define MCP23XXX_ADDR_END` is inside `#ifdef USE_MCP23XXX_DRV` 

#if defined(USE_I2C) || defined(USE_SPI)
#ifdef USE_MCP23XXX_DRV
...
#define MCP23XXX_ADDR_START      0x20     // 32
#define MCP23XXX_ADDR_END        0x26     // 38
....

i understand that `#define USE_MCP23XXX_DRV` means mode 2, so i thought the ADDR parameter even effect in mode 2.

**Reply by sfromis** [reply]
What I meant by "wrapped" was very specifically the symbol itself
> not wrapped in `#ifndef` / `#endif`

That is not the same as just any `#ifdef` for some other symbol.

Basically, you cannot do a new `#define` if the symbol was already defined elsewhere, hence `#ifndef` for if-not-defined.

**Comment by Noschvie** [comment]
https://github.com/arendst/Tasmota/discussions/21244#discussioncomment-11055863

**Reply by ChrisVonFrosta** [reply]
got it. thank you!

**Comment by Noschvie** [comment]
Would be willing to change and test the driver to enable the start and end addresses as build time parameters for Mode 2.  What should be done if necessary?

**Reply by barbudor** [reply]
```
#ifndef xxxx
#define xxxx defaultvalue
#endif
```

**Reply by Noschvie** [reply]
Changed `xdrv_67_mcp23xxx.ino` to

```
#ifndef MCP23XXX_ADDR_START
#define MCP23XXX_ADDR_START      0x20     // 32
#endif
#ifndef MCP23XXX_ADDR_END
#define MCP23XXX_ADDR_END        0x26     // 38
#endif
```
and `user_config_override.h` to

```
#define USE_MCP23XXX_DRV                         // To enable Mode 2
#define MCP23XXX_ADDR_START      0x23            // Overwrite I2C address to start
#define MCP23XXX_ADDR_END        0x28            // Overwrite I2C end address
```

but the "overwrite" doesn't work. 

```
00:00:00.070 Nosch I2C: MCP23XXX_ADDR_START 20
00:00:00.071 Nosch I2C:   MCP23XXX_ADDR_END 26

```
Any idea? Thanks!

**Reply by ChrisVonFrosta** [reply]
as i understood sfromis, those defines doesn't matter for mode 2. 
You have to change the address of the module

**Reply by sfromis** [reply]
My point was that without the driver accommodating overrides, it was not useful try have such `#define`, which are **not** even among the ones suggested by the lines in `my_user_config.h`

**Reply by barbudor** [reply]
What do you mean @sfromis ?
Isn't that accomodating the define ?
```
Line638:     while ((Mcp23x.max_devices < MCP23XXX_MAX_DEVICES) && (mcp23xxx_address < MCP23XXX_ADDR_END)) {
```

**Reply by sfromis** [reply]
In the standard version of the driver, it cannot accommodate an **override**, as the driver itself does the `#define` without it being wrapped in `#ifndef` / `#endif`. And nowhere is it even suggested to try with an override, otherwise I'd expect to see it as a line in `my_user_config.h`

**Reply by barbudor** [reply]
It is the purpose of the lines above by @Noschvie to add such capability in the driver

**Reply by sfromis** [reply]
Right, I was merely trying to clarify what ChrisVonFrosta quoted me about "those defines", referring to a response from very early in this thread.

**Reply by barbudor** [reply]
@Noschvie 
I don't know what you did for it not to work but I changed the driver as here:
https://github.com/barbudor/Tasmota/blob/06f6718338c09907e3ff427dfcd8f3da908b12e4/tasmota/tasmota_xdrv_driver/xdrv_67_mcp23xxx.ino#L78-L83

Adding also a log line:
https://github.com/barbudor/Tasmota/blob/06f6718338c09907e3ff427dfcd8f3da908b12e4/tasmota/tasmota_xdrv_driver/xdrv_67_mcp23xxx.ino#L643

And my `user_config.h`
```
#define USE_MCP23XXX_DRV
#define MCP23XXX_ADDR_START 0x22
#define MCP23XXX_ADDR_END 0x28
```

Got those logs:
```
00:00:00.048 I2C: Bus1 using GPIO03(SCL) and GPIO04(SDA)
00:00:00.051 TFS: File 'mcp23x.dat' not found
00:00:00.052 MCP: scan @22 (34)
00:00:00.053 MCP: scan @23 (35)
00:00:00.053 MCP: scan @24 (36)
00:00:00.064 MCP: scan @25 (37)
00:00:00.065 MCP: scan @26 (38)
00:00:00.065 MCP: scan @27 (39)
```

So it's working fine

Please try again or I'll PR before you 😜 🤣

**Reply by ChrisVonFrosta** [reply]
```
#define USE_MCP23XXX_DRV
#define MCP23XXX_ADDR_START 0x20
#define MCP23XXX_ADDR_END 0x28
```

do i need the address part to get two MCP23017 running?
i have compiled with:
```
#define USE_MCP23XXX_DRV
``` 
but it find only one MCP23017 if both are connected.
If only one is connected, tasmota found it. Whever the module the address is 0x20 or 0x21.

`mcp23x.dat`:
```
{"NAME":"MCP23017 A=B1-8, B=B9-16","GPIO":[32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47]}
{"NAME":"MCP23017 A=B17-24, B=B25-32","GPIO":[48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63]}
```

```
00:00:00.250 CMD: Fall back to serial port, no SOF packet detected on USB port
00:00:00.251 HDW: ESP32-S3 v0.2 (PSRAM)
00:00:00.294 UFS: FlashFS mounted with 12496 kB free
00:00:00.304 CFG: Loaded from File, Count 101
00:00:00.309 QPC: Count 1
00:00:00.310 SPI: Bus1 using GPIO12(CLK), GPIO11(MOSI) and GPIO13(MISO)
00:00:00.311 I2C: Bus1 using GPIO20(SCL) and GPIO19(SDA)
00:00:00.934 I2C: MCP23017 found at 0x20
00:00:00.044 BRY: Berry initialized, RAM used 6578 bytes
00:00:00.063 DSP: File descriptor used
00:00:02.760 DSP: ST7262_1 initialized
00:00:02.768 Project tasmota - TasmotaSunton4.3 Version 14.3.0.4
00:00:02.776 LVG: LVGL initialized
00:00:02.926 I2C: PCA9685 found at 0x40
...
```

**Reply by arendst** [reply]
With more MCP23017 connected check I2C termination. Using the bare chips you definitly need to add resistors.

Then your mp23x.dat file should contain:
```
{"NAME":"MCP23017 A=B1-8, B=B9-16, C=B17-24, D=B25-32","GPIO":[32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63]}
```
Tasmota doesn't care about chips. It just starts using MCP23xxx chips in I2C order until it has found the needed amount of ports. In your case 32 so it uses two chips (if it can see them).

In my case with 32 relais it shows this on restart:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.014 UFS: FlashFS mounted with 1976 kB free
00:00:00.076-028 CFG: Loaded from File, Count 244
00:00:00.082-028 QPC: Count 1
00:00:00.087-028 I2C: Bus1 using GPIO05(SCL) and GPIO04(SDA)
00:00:00.096-028 I2C: MCP23017 found at 0x20
00:00:00.099-028 I2C: MCP23017 found at 0x21
00:00:00.110-028 MCP: Base 0, Template 'MCP23017 A=Ri1-8, B=Ri9-16, C=Ri17-24, D=Ri25-32'
00:00:00.138-028 MCP: Max reached (S0/B0/R32)
00:00:00.138-028 APP: Max number of devices reached
00:00:00.142-028 ROT: Mode 1
00:00:00.144-028 SRC: Restart
00:00:00.158-028 Project tasmota - Wemos14 Version 14.3.0.7(theo)-2_7_8(2024-11-21T16:21:07)
00:00:00.413-028 WIF: Attempting connection...
00:00:01.001-028 WIF: Connecting to AP1 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as wemos14...
00:00:02.421-027 WIF: Connected
00:00:02.674-024 HTP: Web server active on wemos14 with IP address 192.168.2.136
00:00:03.234-023 RTC: UTC 2024-11-22T20:14:12Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
21:14:12.000-023 RTC: Synced by NTP
21:14:12.477-023 MQT: Attempting connection...
21:14:12.492-023 MQT: Connected
21:14:12.496-020 MQT: tele/wemos14/LWT = Online (retained)
21:14:12.498-020 MQT: cmnd/wemos14/POWER = 
21:14:12.500-020 MQT: Subscribe to cmnd/wemos14/#
21:14:12.502-020 MQT: Subscribe to cmnd/displays/#
21:14:12.504-020 MQT: Subscribe to cmnd/DVES_4765B4_fb/#
21:14:12.510-020 MQT: Subscribe to domoticz/out/#
21:14:12.517-020 MQT: tele/wemos14/INFO1 = {"Info1":{"Module":"Generic","Version":"14.3.0.7(theo)","FallbackTopic":"cmnd/DVES_4765B4_fb/","GroupTopic":"cmnd/displays/"}}
21:14:12.530-020 MQT: tele/wemos14/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"wemos14","IPAddress":"192.168.2.136"}}
21:14:12.541-020 MQT: tele/wemos14/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":52}}
21:14:12.552-021 MQT: stat/wemos14/RESULT = {"POWER1":"OFF"}
21:14:12.558-021 MQT: stat/wemos14/POWER1 = OFF
21:14:12.561-020 MQT: stat/wemos14/RESULT = {"POWER2":"OFF"}
21:14:12.565-020 MQT: stat/wemos14/POWER2 = OFF
21:14:12.570-020 MQT: stat/wemos14/RESULT = {"POWER3":"OFF"}
21:14:12.575-020 MQT: stat/wemos14/POWER3 = OFF
21:14:12.579-020 MQT: stat/wemos14/RESULT = {"POWER4":"OFF"}
21:14:12.585-020 MQT: stat/wemos14/POWER4 = OFF
21:14:12.589-020 MQT: stat/wemos14/RESULT = {"POWER5":"OFF"}
21:14:12.594-020 MQT: stat/wemos14/POWER5 = OFF
21:14:12.599-020 MQT: stat/wemos14/RESULT = {"POWER6":"OFF"}
21:14:12.605-021 MQT: stat/wemos14/POWER6 = OFF
21:14:12.611-021 MQT: stat/wemos14/RESULT = {"POWER7":"OFF"}
21:14:12.615-020 MQT: stat/wemos14/POWER7 = OFF
21:14:12.618-020 MQT: stat/wemos14/RESULT = {"POWER8":"ON"}
21:14:12.623-020 MQT: stat/wemos14/POWER8 = ON
21:14:12.627-020 MQT: stat/wemos14/RESULT = {"POWER9":"OFF"}
21:14:12.633-020 MQT: stat/wemos14/POWER9 = OFF
21:14:12.637-020 MQT: stat/wemos14/RESULT = {"POWER10":"ON"}
21:14:12.642-020 MQT: stat/wemos14/POWER10 = ON
21:14:12.647-020 MQT: stat/wemos14/RESULT = {"POWER11":"OFF"}
21:14:12.652-020 MQT: stat/wemos14/POWER11 = OFF
21:14:12.658-021 MQT: stat/wemos14/RESULT = {"POWER12":"OFF"}
21:14:12.664-021 MQT: stat/wemos14/POWER12 = OFF
21:14:12.668-021 MQT: stat/wemos14/RESULT = {"POWER13":"OFF"}
21:14:12.674-021 MQT: stat/wemos14/POWER13 = OFF
21:14:12.678-021 MQT: stat/wemos14/RESULT = {"POWER14":"OFF"}
21:14:12.683-021 MQT: stat/wemos14/POWER14 = OFF
21:14:12.688-021 MQT: stat/wemos14/RESULT = {"POWER15":"OFF"}
21:14:12.693-021 MQT: stat/wemos14/POWER15 = OFF
21:14:12.698-021 MQT: stat/wemos14/RESULT = {"POWER16":"OFF"}
21:14:12.703-021 MQT: stat/wemos14/POWER16 = OFF
21:14:12.708-021 MQT: stat/wemos14/RESULT = {"POWER17":"OFF"}
21:14:12.714-021 MQT: stat/wemos14/POWER17 = OFF
21:14:12.719-021 MQT: stat/wemos14/RESULT = {"POWER18":"OFF"}
21:14:12.724-021 MQT: stat/wemos14/POWER18 = OFF
21:14:12.729-021 MQT: stat/wemos14/RESULT = {"POWER19":"ON"}
21:14:12.734-021 MQT: stat/wemos14/POWER19 = ON
21:14:12.739-021 MQT: stat/wemos14/RESULT = {"POWER20":"OFF"}
21:14:12.744-021 MQT: stat/wemos14/POWER20 = OFF
21:14:12.749-021 MQT: stat/wemos14/RESULT = {"POWER21":"OFF"}
21:14:12.754-021 MQT: stat/wemos14/POWER21 = OFF
21:14:12.760-021 MQT: stat/wemos14/RESULT = {"POWER22":"OFF"}
21:14:12.765-021 MQT: stat/wemos14/POWER22 = OFF
21:14:13.001-021 MQT: stat/wemos14/RESULT = {"POWER23":"OFF"}
21:14:13.006-021 MQT: stat/wemos14/POWER23 = OFF
21:14:13.011-021 MQT: stat/wemos14/RESULT = {"POWER24":"OFF"}
21:14:13.017-021 MQT: stat/wemos14/POWER24 = OFF
21:14:13.022-021 MQT: stat/wemos14/RESULT = {"POWER25":"OFF"}
21:14:13.027-021 MQT: stat/wemos14/POWER25 = OFF
21:14:13.032-021 MQT: stat/wemos14/RESULT = {"POWER26":"OFF"}
21:14:13.037-021 MQT: stat/wemos14/POWER26 = OFF
21:14:13.042-021 MQT: stat/wemos14/RESULT = {"POWER27":"OFF"}
21:14:13.048-021 MQT: stat/wemos14/POWER27 = OFF
21:14:13.053-021 MQT: stat/wemos14/RESULT = {"POWER28":"OFF"}
21:14:13.058-021 MQT: stat/wemos14/POWER28 = OFF
21:14:13.063-021 MQT: stat/wemos14/RESULT = {"POWER29":"OFF"}
21:14:13.068-021 MQT: stat/wemos14/POWER29 = OFF
21:14:13.073-021 MQT: stat/wemos14/RESULT = {"POWER30":"OFF"}
21:14:13.079-021 MQT: stat/wemos14/POWER30 = OFF
21:14:13.084-021 MQT: stat/wemos14/RESULT = {"POWER31":"OFF"}
21:14:13.089-021 MQT: stat/wemos14/POWER31 = OFF
21:14:13.094-021 MQT: stat/wemos14/RESULT = {"POWER32":"OFF"}
21:14:13.099-021 MQT: stat/wemos14/POWER32 = OFF
21:14:14.255-020 MQT: domoticz/in = {"idx":403,"nvalue":0,"svalue":"","Battery":100,"RSSI":10}
21:14:15.256-023 QPC: Reset
21:14:17.217-023 APP: Boot Count 52

```

**Reply by ChrisVonFrosta** [reply]
Thank you!! My mcp23x.dat syntax was the problem

**Comment by Noschvie** [comment]
Hi @barbudor your changes are working, thanks! looks like I had an error in user_config_override.h

**Comment by ChrisVonFrosta** [comment]
just for clearance: i could use more than one MCP23017 (the same to PCF8574) even in Mode1? The modes 2 change only the behavior of its inputs/Outputs to nativ style?

What is about PCA9685? I learned its Mode2 just to be able to connect more than one module

**Reply by sfromis** [reply]
Answer is in the docs: https://tasmota.github.io/docs/MCP230xx/#supporting-modes
But mode 2 is not just to be able to use more than one, it is mainly because it provides a better way of using the expander(s).

There is no "mode 2" for PCA9685, but the V2 driver does seem to have better features than the older, not merely to support more than one chip.

**Reply by ChrisVonFrosta** [reply]
thank you!
